### DISCUSSION TITLE
Composable Multipeer Connectivity

### DISCUSSION THREAD

**Discussion by oliverepper** [pointfreeco/swift-composable-architecture#593] [Ideas]
Hi all!

I recently integrated the Multipeer Connectivity framework into a TCA app. Is there any broader interest for a library like composable-core-location or composable-core-motion? In that case I'd be happy to put some more work in. What I have now is rather minimal and just serves our own needs.

**Comment by jagreenwood** [comment]
I've got a [similar case](https://github.com/jagreenwood/composable-purchase-manager) for in app purchases. It works well and suits my needs and could definitely be built out with more features. I hope to find the time to start working on that soon.
I think the more feature packages we have the better!

**Comment by oliverepper** [comment]
WIP: https://github.com/oliverepper/composable-multipeer-connectivity

It's not done and it's still rough but it serves as a demo. You can advertise, browse and send and receive an invite.

Xcode beta 13.0

✉️ Received Invitation from <MCPeerID: 0x600000255ab0 DisplayName = mona.local-dev>

**Reply by nikitamounier** [reply]
This looks great! What's the advantage of using MultipeerConnectivity over Network.framework, which can also be used for peer-to-peer as shown in [this library](https://github.com/dobster/P2PShareKit)? I was thinking of making my own composable-p2p-network wrapper. I feel like Network.framework is more maintained, more performant, and offers more customisation points, but I'd love to hear your thoughts.

**Reply by oliverepper** [reply]
Oh cool! I simply wasn't aware of other options. I'll check into this. I'd even be interested in something that could work on a Raspberry Pi, too. If maybe TCA can give up the Combine dependency in the future this would be interesting. If you want to start composable-p2p-network wrapper I'd be happy to check it out.

**Reply by nikitamounier** [reply]
I've started implementing [composable-P2P](https://github.com/nikitamounier/composable-P2P), I'd love for you to check it out! I've created the interfaces for the Listener, Browser, and Connection (check out the [Network.framework documentation](https://developer.apple.com/documentation/network) for more context), and I also implemented the Type-Length-Value protocol. I'm just not too sure about the structure composable-P2P should take – the Connection type is kind of like your Session type, but not exactly. Let me know what you think.

**Comment by nikitamounier** [comment]
I've implemented [composable-P2P](https://github.com/nikitamounier/composable-P2P) which uses the Network framework instead of MultipeerConnectivity, let me know what you think. Had to do some pretty creative stuff to make it work with TCA's composable dependancy pattern 😅
